### PR TITLE
Completed: Dark theme for FF Notes

### LINKED ISSUE #181: Add Dark Theme to Notes
Many users prefer dark themes for their browsers as it's easier on the eyes than light themes when using your computer for long hours.

Could we add a dark theme option to Firefox Notes?

### PR DESCRIPTION
**Pull request by cedricium**
Fixes #181 

Eventual fix for #181. This is very much still a WIP and any input is much appreciated.

## How it Currently Works:
There are two CSS files, `styles.css` and `styles-dark.css`, corresponding to the Default (light) and Dark themes, respectively.

To switch themes users go into the Settings page and select the corresponding radio button and hit 'Save', which stores their preference in `localStorage`. When the notepad is initially loaded, it looks for the saved theme preference in `localStorage` and sets the appropriate CSS file as the href link in the `<link id="main-css" ...>` tag in `panel.html`.

Because the theme switching only takes place when the notepad is initially loaded (on `DOMContentLoaded`), users must close then reopen the notepad for the theme-switching to take place.

## Preview of Theme:
screen shot 2017-08-04 at 12 25 57 am

The styling is based off of the Firefox Devtools Dark theme (as close as I could get). I know because this is an official Mozilla product the colors/theme will need to be to spec (@ryanfeeley mind helping out?).

## Preview of Settings Page:
screen shot 2017-08-04 at 12 31 26 am

## What's Planned:

- [x] theme updating **as soon as** new preference is chosen (WIP with `background.js`) (**updated 8/04** - dd0a562)
- [x] refined styling of Dark theme (**updated 8/05** - 5b2e4b1)
- [x] adding error handling and setting the default theme (**updated 8/04** - dd0a562)
- [x] making the Settings page a little nicer (**updated 8/05** - a54e970)
- [x] set up localization for the Settings page (**updated 8/05** - 1cb9d3a)

### CODE REVIEW

**Comment by cedricium** [comment]
## Update - Immediate Theme Change upon Saving:

notes_theme-switching

**Comment by cedricium** [comment]
### Settings page styling
I'm curious as to how the Settings page should be styled. This is how it currently is set up in this PR:
notes_settings

But because this is an official Mozilla project I'm assuming it should follow a more standard guideline, something like Firefox's own Options page:
image

**Comment by ryanfeeley** [comment]
@cedricium The recommendations for styling add-on preferences is still under development:
https://developer.mozilla.org/en-US/Add-ons/WebExtensions/user_interface/Options_pages

And related background:
https://bugzilla.mozilla.org/show_bug.cgi?id=1369209

My advice? Personally I think the more you can make it look like the "Automatic Updates" row, the better. Like this but horizontal:

 Theme
- [x] Default
- [ ] Dark

Using "Light" may not even be necessary as users can safely assume that something that isn't dark is likely lighter.

**Comment by vladikoff** [comment]
@cedricium we just tested it and it looks almost ready to merge!

Some things we noticed: 

1. When we `npm start` the extension and the sidebar shows there is a flash of the DARK theme background. It is very subtle but you can probably notice it.

2. In `options` can we void using the `Save` button and just make the style switch when the radio button changes?

3. Can we make the highlight match the DevTools Highlight:

Notes: 

image

DevTools:

image

**Comment by cedricium** [comment]
@ryanfeeley thanks for the advice, I'll run with mimicking the "Automatic Updates" section and change the "Light" label to "Default".

And @vladikoff, good to hear. I'll see what I can do about those issues 👍

**Comment by cedricium** [comment]
@vladikoff Everything you and @ryanfeeley mentioned has been resolved except for the flashing of the dark theme on the initial page load. I can't seem to find a solution: I tried adding the stylesheet dynamically after the page has been loaded, found out about [FOUC](https://en.wikipedia.org/wiki/Flash_of_unstyled_content) and tried using various techniques to solve that and those didn't work.

Do you have any ideas on how to fix it?

**Comment by Natim** [comment]
@vladikoff I will let you merge if it is good for you with the changes.

**Comment by cedricium** [comment]
I forgot to actually change the 'Light' label localized string to 'Default', it's now done. Everything should be good to go besides the flickering on initial load.

**Comment by vladikoff** [comment]
@cedricium To avoid the Dark Theme flash try something like this, you might be able to clean up a bit:  (It seems like `disabled` attribute is not the way to go with this)

```diff
diff --git a/src/sidebar/panel.html b/src/sidebar/panel.html
index 9ef4b5a..c66e420 100644
--- a/src/sidebar/panel.html
+++ b/src/sidebar/panel.html
@@ -5,7 +5,6 @@
   <title>Firefox Notes</title>
   <link href="vendor/quill.snow.css" rel="stylesheet">
   <link id="main-css" href="styles.css" rel="stylesheet">
-  <link id="dark-css" href="styles-dark.css" rel="stylesheet" disabled>
 </head>
 
 <body>
diff --git a/src/sidebar/panel.js b/src/sidebar/panel.js
index afdbc82..cbdb90e 100644
--- a/src/sidebar/panel.js
+++ b/src/sidebar/panel.js
@@ -184,9 +184,18 @@ function getThemeFromStorage() {
     const darkCSS = document.getElementById('dark-css');
     
     if (data.theme === 'dark') {
-      darkCSS.disabled = false;
+      var ss = document.createElement("link");
+      ss.id = 'dark-styles';
+      ss.type = "text/css";
+      ss.rel = "stylesheet";
+      ss.href = "styles-dark.css";
+      document.getElementsByTagName("head")[0].appendChild(ss);
+
+      //darkCSS.disabled = false;
     } else if (data.theme === 'default' || data.theme === undefined) {
-      darkCSS.disabled = true;
+      const elem = document.getElementById("dark-styles");
+      elem.parentElement.removeChild(elem);
+      //darkCSS.disabled = true;
     }
   });
 }

```

File: src/manifest.json [hunk 1]
```diff
@@ -32,7 +32,11 @@
       "background.js"
     ]
   },
-
+
```
Comment (vladikoff) [inline src/manifest.json]: we don't have new lines between other properties, so we can remove these

File: src/settings/settings.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+for (let i = 0; i < themeRadioBtn.length; i++) {
+  themeRadioBtn[i].onclick = function() {
+    const theme = getTheme();
+  
+    browser.storage.local.set(theme);
+
+    // notify background.js that theme settings have changed
+    browser.runtime.sendMessage({
+      action: 'theme-changed'
```
Comment (vladikoff) [inline src/settings/settings.js]: @cedricium Can we add a metrics event if users change themes? You should be able to find how other metrics are sent in the codebase. Let me know if you need help

Comment (cedricium) [inline src/settings/settings.js]: Hey @vladikoff, thanks for taking a look at this PR, I appreciate all your help.

As for the metrics, I understand I need to declare either a custom metric or customer dimension per the [metrics.md](https://github.com/mozilla/notes/blob/master/docs/metrics.md) doc, but I'm not sure where I would define that. I'm confused too as to how to debug sending metrics/making sure they are sent.

Comment (vladikoff) [inline src/settings/settings.js]: @cedricium Just call `sendMetrics` from background scripts.

https://github.com/mozilla/notes/blob/master/src/background.js#L44

`sendMetrics('theme-changed', eventData.context);`

Does that help?

Comment (vladikoff) [inline src/settings/settings.js]: Oh also add that event into metrics.md file

Comment (cedricium) [inline src/settings/settings.js]: @vladikoff just pushed the changes, let me know if there's anything else I need.

File: src/background.js [hunk 3]
```diff
@@ -51,6 +51,7 @@ browser.runtime.onMessage.addListener(function(eventData) {
       sendMetrics('drag-n-drop', eventData.context);
       break;
     case 'theme-changed':
+      sendMetrics('theme-changed', eventData.content);
```
Comment (vladikoff) [inline src/background.js]: 👍

File: src/sidebar/panel.html [hunk 4]
```diff
@@ -5,7 +5,7 @@
   <title>Firefox Notes</title>
   <link href="vendor/quill.snow.css" rel="stylesheet">
   <link id="main-css" href="styles.css" rel="stylesheet">
-  <link id="dark-css" href="styles-dark.css" rel="stylesheet" disabled>
+<!--  <link id="dark-css" href="styles-dark.css" rel="stylesheet" disabled>-->
```
Comment (vladikoff) [inline src/sidebar/panel.html]: @cedricium we can just remove this. not comment it out

File: src/sidebar/panel.js [hunk 5]
```diff
@@ -180,13 +180,23 @@ enableSync.onclick = () => {
 // file to the document
 function getThemeFromStorage() {
   const getting = browser.storage.local.get(['theme']);
-  getting.then(function applyTheme(data) {    
-    const darkCSS = document.getElementById('dark-css');
-    
+  getting.then(function applyTheme(data) {
     if (data.theme === 'dark') {
-      darkCSS.disabled = false;
+      if (!document.getElementById('dark-styles')) {
```
Comment (vladikoff) [inline src/sidebar/panel.js]: style nit: add a space between `!` and `document`

File: src/sidebar/panel.html [hunk 6]
```diff
@@ -4,7 +4,8 @@
   <meta charset="utf-8" />
   <title>Firefox Notes</title>
   <link href="vendor/quill.snow.css" rel="stylesheet">
-  <link href="styles.css" rel="stylesheet">
+  <link id="main-css" href="styles.css" rel="stylesheet">
+<!--  <link id="dark-css" href="styles-dark.css" rel="stylesheet" disabled>-->
```
Comment (vladikoff) [inline src/sidebar/panel.html]: @cedricium let's remove this

**Comment by vladikoff** [comment]
@cedricium users are already excited https://www.reddit.com/r/firefox/comments/6ujvwy/firefox_notes_now_has_a_dark_theme/ . ! :)

**Comment by cedricium** [comment]
@vladikoff That is so cool! Such an awesome feeling knowing people are using something you've made :smiley:

**Pull request merged**